internal/pkg/cosign/sign.go

```diff
@@ -0,0 +1,188 @@
+}
+
+// Sign uses the PayloadSigner to sign the requested payload, then returns the signature, the public key associated with it, the signed payload
+func (ps *PayloadSigner) Sign(ctx context.Context, req *SigningRequest) (*SigningResults, error) {
+	sOpts := []signature.SignOption{signatureoptions.WithContext(ctx)}
+	sOpts = append(sOpts, ps.PayloadSignerOpts...)
+	sig, err := ps.PayloadSigner.SignMessage(bytes.NewReader(req.SignaturePayload), sOpts...)
+	if err != nil {
+		return nil, err
+	}
+
+	pkOpts := []signature.PublicKeyOption{signatureoptions.WithContext(ctx)}
+	pkOpts = append(pkOpts, ps.PublicKeyProviderOpts...)
+	pk, err := ps.PayloadSigner.PublicKey(pkOpts...)
+	if err != nil {
+		return nil, err
+	}
+
+	return &SigningResults{
+		Signature:     sig,
+		SignedPayload: req.SignaturePayload,
+		SignedEntity:  req.SignedEntity,
+		Pub:           pk,
+	}, nil
+}
+
+// RemoteSignerWrapper implements `Signer`
+type FulcioSignerWrapper struct {
+	Inner Signer
+
+	Cert, Chain []byte
+}
+
+func (fs *FulcioSignerWrapper) Sign(ctx context.Context, req *SigningRequest) (*SigningResults, error) {
+	results, err := fs.Inner.Sign(ctx, req)
+	if err != nil {
+		return nil, err
+	}
+
+	// TODO(dekkagaijin): move the fulcio SignerVerififer logic here
+
+	results.Cert = fs.Cert
+	results.Chain = fs.Chain
+
+	return results, nil
+}
+
+// RemoteSignerWrapper implements `Signer`
+type OCISignatureBuilder struct {
+	Inner Signer
+}
+
+// Sign calls the Inner signer and then returns the `oci.Signature` corresponding to the Signature, Cert, Chain, and Bundle in the overall result
+func (sb *OCISignatureBuilder) Sign(ctx context.Context, req *SigningRequest) (*SigningResults, error) {
+	results, err := sb.Inner.Sign(ctx, req)
+	if err != nil {
+		return nil, err
+	}
+
+	opts := []static.Option{}
+	if results.Cert != nil {
+		opts = append(opts, static.WithCertChain(results.Cert, results.Chain))
+	}
+	if results.Bundle != nil {
+		opts = append(opts, static.WithBundle(results.Bundle))
+	}
+	b64sig := base64.StdEncoding.EncodeToString(results.Signature)
+
+	// Create the new signature for this entity.
+	ociSig, err := static.NewSignature(results.SignedPayload, b64sig, opts...)
+	if err != nil {
+		return nil, err
+	}
+	results.OCISignature = ociSig
+
+	return results, nil
+}
+
+// RemoteSignerWrapper implements `Signer`
+type OCISignatureAttacher struct {
+	Inner Signer
+
+	DD mutate.DupeDetector
+}
+
+// Sign calls the Inner signer and then attaches the resulting `oci.Signature` to the resulting `oci.SignedEntity`
+func (sa *OCISignatureAttacher) Sign(ctx context.Context, req *SigningRequest) (*SigningResults, error) {
+	results, err := sa.Inner.Sign(ctx, req)
+	if err != nil {
+		return nil, err
+	}
+
+	// Attach the signature to the entity.
+	newSE, err := mutate.AttachSignatureToEntity(results.SignedEntity, results.OCISignature, mutate.WithDupeDetector(sa.DD))
+	if err != nil {
+		return nil, err
+	}
+	results.SignedEntity = newSE
```

Comment:
Don't conflate signing with mutating the signed entity to carry that signature.  The signer should *just* sign, let the caller worry about this.

Reply:
roger